python/hail/typecheck.py

```diff
@@ -0,0 +1,123 @@
+from types import ClassType
+import re
+
+
+def extract(t):
+    m = re.match("<(type|class) '(.*)'>", str(t))
+    if m:
+        return m.groups()[1]
+    else:
+        return str(t)
+
+
+class TypeChecker(object):
+    def __init__(self, f, supported):
+        self.f = f
+        self.supported = supported
+
+    def check(self, x):
+        return self.f(x)
+
+
+class BaseChecker(TypeChecker):
+    def __init__(self, f, s):
+        super(BaseChecker, self).__init__(lambda x: f(x), s)
+
+
+class MultipleTypeChecker(TypeChecker):
+    def __init__(self, checkers):
+        def check(x):
+            any_pass = False
+            for c in checkers:
+                any_pass = any_pass or c.check(x)
+            return any_pass
+
+        supported = '[ ' + ' or '.join([c.supported for c in checkers]) + ' ]'
+
+        super(MultipleTypeChecker, self).__init__(check, supported)
+
+
+class CollectionChecker(TypeChecker):
+    def __init__(self, collection_checker, element_checker):
+        def check(x):
+            passes = collection_checker.check(x)
+            if passes:
+                for elt in x:
+                    passes = passes and element_checker.check(elt)
+            return passes
+
+        s = '(%s of %s)' % (collection_checker.supported, element_checker.supported)
+        super(CollectionChecker, self).__init__(check, s)
+
+
+class DictChecker(TypeChecker):
+    def __init__(self, key_checker, value_checker):
+
+        def check(x):
+            passes = isinstance(x, dict)
+            if passes:
+                for k, v in x.iteritems():
+                    passes = passes and key_checker.check(k)
+                    passes = passes and value_checker.check(v)
+            return passes
+
+        s = '(dict of %s, %s)' % (key_checker.supported, value_checker.supported)
+        super(DictChecker, self).__init__(check, s)
+
+
+null = BaseChecker(lambda x: x is None, 'None')
+
+any = BaseChecker(lambda x: True, 'Any')
+
+
+def only(t):
+    if isinstance(t, type) or type(t) is ClassType:
+        return BaseChecker(lambda x: isinstance(x, t), extract(t))
+    elif isinstance(t, TypeChecker):
+        return t
+    else:
+        raise RuntimeError("invalid typecheck signature: expected 'type' or 'TypeChecker', found '%s'" % type(t))
+
+
+def oneof(*args):
+    return MultipleTypeChecker([only(x) for x in args])
+
+
+def nullable(t):
+    return oneof(only(t), null)
+
+
+def listof(t):
+    return CollectionChecker(only(list), only(t))
+
+
+def dictof(k, v):
+    return DictChecker(only(k), only(v))
+
+
+strlike = oneof(str, unicode)
+
+integral = oneof(int, long)
+
+numeric = oneof(int, long, float)
+
+
+def typecheck(**kw):
+
+    def _typecheck(f, *args, **kwargs):
+        argspec = getargspec(f).args
+        for kwarg in kw.keys():
+            if not kwarg in argspec:
+                raise RuntimeError("invalid typecheck signature: unmatched parameter '%s'" % kwarg)
+        for i, arg in enumerate(args):
```

Comment:
It doesn't look like your are type checking the keyword arguments here.  Only the  positional arguments appear in `args`:

```
>>> def f(*args, **kwargs):
...   return (args, kwargs)
... 
>>> f(1, 2, k=5)
((1, 2), {'k': 5})
```